start.go

```diff
@@ -144,3 +144,13 @@ func UpdateFilter(allow, block []string) {
 func UpdateFilterAndLabels(allow, block, allowedLabels, blockedLabels []string) {
 	globalMetrics.Load().(*Metrics).UpdateFilterAndLabels(allow, block, allowedLabels, blockedLabels)
 }
+
+// Shutdown flushes and disables metric collection, blocking while waiting for this to complete.
+// WARNING: Not all MetricSink backends support this functionality, and calling this will cause resource leaks.
+// This is intended for use immediately prior to application exit.
+func Shutdown() {
+	m := globalMetrics.Load().(*Metrics)
+	// Replace global metrics with the BlackholeSink like how init setup the library.
+	globalMetrics.Store(&Metrics{sink: &BlackholeSink{}})
```

Comment:
I don't understand this line 100% <-- would you mind helping me understand why we need to do this?

Reply:
The goal is to explicitly break the following example, so that library consumers build their applications knowing that nothing is collected post `Shutdown`, and don't accidentally become dependant on collecting metrics post-`Shutdown`.
```
import "github.com/armon/go-metrics"

func main() {
  // The implementation of the metrics library means
  // that {"hello", "world"} will only ever have the value 1.
  metrics.New(...)
  metrics.IncrCounter([]string{"hello", "world"}, 1)
  metrics.Shutdown()
  metrics.IncrCounter([]string{"hello", "world"}, 1)
}
```

Resetting the library to the `BlackholeSink` isn't strictly necessary. But between applications not cleaning up goroutines (that generate calls to the library), and `MetricSink` implementations not supporting shutdown/close, it's possible that applications collect metrics and then upload them in the small window between `Shutdown` and exit. This makes it pretty explicit that this isn't intended or supported.

There is a race condition here: callers getting a pointer to the underlying `Metrics` struct and then completing their call into the `MetricSink` after `Shutdown` finishes. `MetricSink` must be thread-safe, so this is likely fine.

Reply:
Thanks for laying out for me here 💯 I see now the issue that could happen; I'll resolve this comment for now.